File: leetcode/java/98.validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public boolean isValidBST(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        traverse(root,list);

        for (int i = 1; i < list.size(); i++) {
            if(list.get(i) <= list.get(i-1)){
                return false;
            }
        }
        return true;
    }

    private static void traverse(TreeNode root, List<Integer> res) {
        if (root == null) {
            return;
        }

        if (isLeaf(root)) {
            res.add(root.val);
            return;
        }

        traverse(root.left, res);
        res.add(root.val);
        traverse(root.right, res);
    }

    private static boolean isLeaf(TreeNode root) {
        return root != null && root.left == null && root.right == null;
    }
}

File: leetcode/cpp/98.validate-binary-search-tree.cpp
#include "oj_header.h"

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include<vector>
using namespace std;

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        if(!root) return true;
        vector<int> res;
        to_vector(res , root);
        for(int i = 0; i < res.size() -1 ; ++i){
            if(res[i] >= res[i + 1])
                return false;
        }
        return true;
    }
    void to_vector(vector<int> &res , TreeNode* root){
        if(!root) return;
        to_vector(res,root->left);
        res.push_back(root->val);
        to_vector(res,root->right);
    }
};
